File: apps/backend/app/api/routes/notifications.py
"""
Notification and Onboarding API Routes
"""
from datetime import datetime
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete, func, and_
from pydantic import BaseModel, Field

from app.core.database import get_db
from app.models.notifications import (
    Notification,
    NotificationSettings,
    UserProfile,
    NotificationType,
    NotificationPriority,
    UserProfileType,
    COMMON_WORK_APPS,
    PROFILE_DESCRIPTIONS,
)


router = APIRouter()


# ============================================================================
# Pydantic Schemas
# ============================================================================

class NotificationAction(BaseModel):
    """Action button for notification"""
    label: str
    action: str  # Action identifier
    primary: bool = False


class NotificationCreate(BaseModel):
    """Create notification request"""
    type: str = NotificationType.INFO.value
    title: str
    message: str
    icon: Optional[str] = None
    actions: Optional[List[NotificationAction]] = None
    priority: str = NotificationPriority.NORMAL.value
    source: Optional[str] = None
    data: Optional[dict] = None
    show_native: bool = False
    expires_at: Optional[datetime] = None


class NotificationResponse(BaseModel):
    """Notification response"""
    id: int
    type: str
    title: str
    message: str
    icon: Optional[str]
    actions: Optional[List[dict]]
    is_read: bool
    is_dismissed: bool
    priority: str
    source: Optional[str]
    data: Optional[dict]
    created_at: datetime
    read_at: Optional[datetime]
    expires_at: Optional[datetime]
    show_native: bool

    class Config:
        from_attributes = True


class NotificationSettingsUpdate(BaseModel):
    """Update notification settings"""
    notifications_enabled: Optional[bool] = None
    distraction_alerts: Optional[bool] = None
    distraction_threshold_minutes: Optional[int] = None
    goal_notifications: Optional[bool] = None
    streak_notifications: Optional[bool] = None
    focus_reminders: Optional[bool] = None
    break_reminders: Optional[bool] = None
    break_threshold_minutes: Optional[int] = None
    daily_summary: Optional[bool] = None
    daily_summary_time: Optional[str] = None
    weekly_report: Optional[bool] = None
    weekly_report_day: Optional[int] = None
    ai_insights: Optional[bool] = None
    quiet_hours_enabled: Optional[bool] = None
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    native_notifications: Optional[bool] = None
    sound_enabled: Optional[bool] = None


class NotificationSettingsResponse(BaseModel):
    """Notification settings response"""
    id: int
    notifications_enabled: bool
    distraction_alerts: bool
    distraction_threshold_minutes: int
    goal_notifications: bool
    streak_notifications: bool
    focus_reminders: bool
    break_reminders: bool
    break_threshold_minutes: int
    daily_summary: bool
    daily_summary_time: str
    weekly_report: bool
    weekly_report_day: int
    ai_insights: bool
    quiet_hours_enabled: bool
    quiet_hours_start: str
    quiet_hours_end: str
    native_notifications: bool
    sound_enabled: bool

    class Config:
        from_attributes = True


class UserProfileUpdate(BaseModel):
    """Update user profile"""
    profile_type: Optional[str] = None
    work_apps: Optional[List[str]] = None
    daily_productive_hours: Optional[float] = None
    max_distraction_hours: Optional[float] = None
    ai_enabled: Optional[bool] = None
    openai_api_key: Optional[str] = None
    accessibility_granted: Optional[bool] = None
    screen_recording_granted: Optional[bool] = None
    launch_on_startup: Optional[bool] = None


class UserProfileResponse(BaseModel):
    """User profile response"""
    id: int
    onboarding_completed: bool
    onboarding_completed_at: Optional[datetime]
    onboarding_step: int
    profile_type: str
    work_apps: List[str]
    daily_productive_hours: float
    max_distraction_hours: float
    ai_enabled: bool
    accessibility_granted: bool
    screen_recording_granted: bool
    launch_on_startup: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class OnboardingStepData(BaseModel):
    """Data for a specific onboarding step"""
    step: int
    data: dict


class CommonAppsResponse(BaseModel):
    """Common apps for onboarding"""
    apps: List[dict]


class UnreadCountResponse(BaseModel):
    """Unread notification count"""
    count: int


# ============================================================================
# Notification Routes
# ============================================================================

@router.get("/notifications", response_model=List[NotificationResponse])
async def list_notifications(
    limit: int = Query(50, ge=1, le=200),
    unread_only: bool = Query(False),
    type: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """List notifications"""
    query = select(Notification).where(Notification.is_dismissed == False)

    if unread_only:
        query = query.where(Notification.is_read == False)

    if type:
        query = query.where(Notification.type == type)

    query = query.order_by(Notification.created_at.desc()).limit(limit)
    result = await db.execute(query)
    notifications = result.scalars().all()

    return notifications


@router.get("/notifications/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(db: AsyncSession = Depends(get_db)):
    """Get unread notification count"""
    query = select(func.count(Notification.id)).where(
        and_(
            Notification.is_read == False,
            Notification.is_dismissed == False
        )
    )
    result = await db.execute(query)
    count = result.scalar() or 0

    return UnreadCountResponse(count=count)


@router.post("/notifications", response_model=NotificationResponse)
async def create_notification(
    notification: NotificationCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new notification"""
    db_notification = Notification(
        type=notification.type,
        title=notification.title,
        message=notification.message,
        icon=notification.icon,
        actions=[a.model_dump() for a in notification.actions] if notification.actions else None,
        priority=notification.priority,
        source=notification.source,
        data=notification.data,
        show_native=notification.show_native,
        expires_at=notification.expires_at,
    )

    db.add(db_notification)
    await db.commit()
    await db.refresh(db_notification)

    return db_notification


@router.put("/notifications/{notification_id}/read", response_model=NotificationResponse)
async def mark_notification_read(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Mark a notification as read"""
    query = select(Notification).where(Notification.id == notification_id)
    result = await db.execute(query)
    notification = result.scalar_one_or_none()

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    notification.is_read = True
    notification.read_at = datetime.utcnow()
    await db.commit()
    await db.refresh(notification)

    return notification


@router.delete("/notifications/{notification_id}")
async def delete_notification(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Delete (dismiss) a notification"""
    query = select(Notification).where(Notification.id == notification_id)
    result = await db.execute(query)
    notification = result.scalar_one_or_none()

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    notification.is_dismissed = True
    await db.commit()

    return {"status": "success", "message": "Notification dismissed"}


@router.post("/notifications/read-all")
async def mark_all_notifications_read(db: AsyncSession = Depends(get_db)):
    """Mark all notifications as read"""
    await db.execute(
        update(Notification)
        .where(Notification.is_read == False)
        .values(is_read=True, read_at=datetime.utcnow())
    )
    await db.commit()

    return {"status": "success", "message": "All notifications marked as read"}


@router.delete("/notifications/clear")
async def clear_all_notifications(db: AsyncSession = Depends(get_db)):
    """Clear all notifications (mark as dismissed)"""
    await db.execute(
        update(Notification)
        .values(is_dismissed=True)
    )
    await db.commit()

    return {"status": "success", "message": "All notifications cleared"}


# ============================================================================
# Notification Settings Routes
# ============================================================================

@router.get("/notifications/settings", response_model=NotificationSettingsResponse)
async def get_notification_settings(db: AsyncSession = Depends(get_db)):
    """Get notification settings"""
    query = select(NotificationSettings)
    result = await db.execute(query)
    settings = result.scalar_one_or_none()

    if not settings:
        # Create default settings
        settings = NotificationSettings()
        db.add(settings)
        await db.commit()
        await db.refresh(settings)

    return settings


@router.put("/notifications/settings", response_model=NotificationSettingsResponse)
async def update_notification_settings(
    updates: NotificationSettingsUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update notification settings"""
    query = select(NotificationSettings)
    result = await db.execute(query)
    settings = result.scalar_one_or_none()

    if not settings:
        settings = NotificationSettings()
        db.add(settings)

    # Update only provided fields
    update_data = updates.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(settings, field, value)

    await db.commit()
    await db.refresh(settings)

    return settings

File: apps/desktop/src/lib/api/notifications.ts
/**
 * Notifications and Onboarding API
 */
import axios from 'axios';

const API_BASE = 'http://localhost:8000/api';

// ============================================================================
// Types
// ============================================================================

export type NotificationType =
  | 'info'
  | 'success'
  | 'warning'
  | 'error'
  | 'ai_insight'
  | 'distraction_alert'
  | 'goal_progress'
  | 'goal_achieved'
  | 'focus_reminder'
  | 'break_suggestion'
  | 'daily_summary'
  | 'streak_alert'
  | 'weekly_report';

export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent';

export type UserProfileType =
  | 'developer'
  | 'designer'
  | 'writer'
  | 'manager'
  | 'student'
  | 'freelancer'
  | 'other';

export interface NotificationAction {
  label: string;
  action: string;
  primary?: boolean;
}

export interface Notification {
  id: number;
  type: NotificationType;
  title: string;
  message: string;
  icon?: string;
  actions?: NotificationAction[];
  is_read: boolean;
  is_dismissed: boolean;
  priority: NotificationPriority;
  source?: string;
  data?: Record<string, unknown>;
  created_at: string;
  read_at?: string;
  expires_at?: string;
  show_native: boolean;
}

export interface NotificationCreate {
  type?: NotificationType;
  title: string;
  message: string;
  icon?: string;
  actions?: NotificationAction[];
  priority?: NotificationPriority;
  source?: string;
  data?: Record<string, unknown>;
  show_native?: boolean;
  expires_at?: string;
}

export interface NotificationSettings {
  id: number;
  notifications_enabled: boolean;
  distraction_alerts: boolean;
  distraction_threshold_minutes: number;
  goal_notifications: boolean;
  streak_notifications: boolean;
  focus_reminders: boolean;
  break_reminders: boolean;
  break_threshold_minutes: number;
  daily_summary: boolean;
  daily_summary_time: string;
  weekly_report: boolean;
  weekly_report_day: number;
  ai_insights: boolean;
  quiet_hours_enabled: boolean;
  quiet_hours_start: string;
  quiet_hours_end: string;
  native_notifications: boolean;
  sound_enabled: boolean;
}

export interface NotificationSettingsUpdate {
  notifications_enabled?: boolean;
  distraction_alerts?: boolean;
  distraction_threshold_minutes?: number;
  goal_notifications?: boolean;
  streak_notifications?: boolean;
  focus_reminders?: boolean;
  break_reminders?: boolean;
  break_threshold_minutes?: number;
  daily_summary?: boolean;
  daily_summary_time?: string;
  weekly_report?: boolean;
  weekly_report_day?: number;
  ai_insights?: boolean;
  quiet_hours_enabled?: boolean;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
  native_notifications?: boolean;
  sound_enabled?: boolean;
}

export interface UserProfile {
  id: number;
  onboarding_completed: boolean;
  onboarding_completed_at?: string;
  onboarding_step: number;
  profile_type: UserProfileType;
  work_apps: string[];
  daily_productive_hours: number;
  max_distraction_hours: number;
  ai_enabled: boolean;
  accessibility_granted: boolean;
  screen_recording_granted: boolean;
  launch_on_startup: boolean;
  created_at: string;
  updated_at: string;
}

export interface UserProfileUpdate {
  profile_type?: UserProfileType;
  work_apps?: string[];
  daily_productive_hours?: number;
  max_distraction_hours?: number;
  ai_enabled?: boolean;
  openai_api_key?: string;
  accessibility_granted?: boolean;
  screen_recording_granted?: boolean;
  launch_on_startup?: boolean;
}

export interface OnboardingStatus {
  completed: boolean;
  step: number;
  profile_exists: boolean;
}

export interface OnboardingStepData {
  step: number;
  data: Record<string, unknown>;
}

export interface CommonApp {
  name: string;
  icon: string;
  category: string;
}

// ============================================================================
// Notification API Functions
// ============================================================================

export async function fetchNotifications(
  limit: number = 50,
  unreadOnly: boolean = false,
  type?: NotificationType
): Promise<Notification[]> {
  const params = new URLSearchParams();
  params.append('limit', limit.toString());
  if (unreadOnly) params.append('unread_only', 'true');
  if (type) params.append('type', type);

  const response = await axios.get<Notification[]>(
    `${API_BASE}/notifications?${params.toString()}`
  );
  return response.data;
}

export async function fetchUnreadCount(): Promise<number> {
  const response = await axios.get<{ count: number }>(
    `${API_BASE}/notifications/unread-count`
  );
  return response.data.count;
}

export async function createNotification(
  notification: NotificationCreate
): Promise<Notification> {
  const response = await axios.post<Notification>(
    `${API_BASE}/notifications`,
    notification
  );
  return response.data;
}

export async function markNotificationRead(id: number): Promise<Notification> {
  const response = await axios.put<Notification>(
    `${API_BASE}/notifications/${id}/read`
  );
  return response.data;
}

export async function deleteNotification(id: number): Promise<void> {
  await axios.delete(`${API_BASE}/notifications/${id}`);
}

export async function markAllNotificationsRead(): Promise<void> {
  await axios.post(`${API_BASE}/notifications/read-all`);
}

export async function clearAllNotifications(): Promise<void> {
  await axios.delete(`${API_BASE}/notifications/clear`);
}

export async function fetchNotificationSettings(): Promise<NotificationSettings> {
  const response = await axios.get<NotificationSettings>(
    `${API_BASE}/notifications/settings`
  );
  return response.data;
}

export async function updateNotificationSettings(
  settings: NotificationSettingsUpdate
): Promise<NotificationSettings> {
  const response = await axios.put<NotificationSettings>(
    `${API_BASE}/notifications/settings`,
    settings
  );
  return response.data;
}

// ============================================================================
// Onboarding API Functions
// ============================================================================

export async function fetchOnboardingStatus(): Promise<OnboardingStatus> {
  const response = await axios.get<OnboardingStatus>(
    `${API_BASE}/onboarding/status`
  );
  return response.data;
}

export async function fetchUserProfile(): Promise<UserProfile> {
  const response = await axios.get<UserProfile>(
    `${API_BASE}/onboarding/profile`
  );
  return response.data;
}

export async function updateUserProfile(
  updates: UserProfileUpdate
): Promise<UserProfile> {
  const response = await axios.put<UserProfile>(
    `${API_BASE}/onboarding/profile`,
    updates
  );
  return response.data;
}

export async function saveOnboardingStep(
  step: number,
  data: Record<string, unknown>
): Promise<void> {
  await axios.post(`${API_BASE}/onboarding/step`, { step, data });
}

export async function completeOnboarding(): Promise<{
  status: string;
  message: string;
  profile: {
    profile_type: string;
    work_apps: string[];
    daily_productive_hours: number;
  };
}> {
  const response = await axios.post(`${API_BASE}/onboarding/complete`);
  return response.data;
}

export async function resetOnboarding(): Promise<void> {
  await axios.post(`${API_BASE}/onboarding/reset`);
}

export async function fetchCommonApps(): Promise<CommonApp[]> {
  const response = await axios.get<{ apps: CommonApp[] }>(
    `${API_BASE}/onboarding/common-apps`
  );
  return response.data.apps;
}

export async function fetchProfileTypes(): Promise<{
  types: Array<{
    value: UserProfileType;
    label: string;
    description: Record<string, unknown>;
  }>;
}> {
  const response = await axios.get(`${API_BASE}/onboarding/profile-types`);
  return response.data;
}
